File: backend/kernel/approval_rules.py
"""审批规则（settings KV「approval_rules」）——真正被内核消费。

与前端 Approvals 页 RulesModal 同源 key：
  auto_low_risk / review_high_risk / review_capability_upgrade /
  review_evolution / auto_tighten_2x

规则只影响「是否自动放行 / 是否必须人工」；
进化建议（evolution）永不自动应用（硬红线，本模块不提供后门）。
"""

from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)

# 最近一次成功加载的规则（sync 路径 / charge_tokens 用）
_RULES_CACHE: list[dict[str, Any]] | None = None

# 默认与前端 DEFAULT_RULES 对齐
DEFAULT_RULES: list[dict[str, Any]] = [
    {"key": "auto_low_risk", "enabled": True},
    {"key": "review_high_risk", "enabled": True, "warn": True},
    {"key": "review_capability_upgrade", "enabled": True, "warn": True},
    {"key": "review_evolution", "enabled": True, "warn": True},
    {"key": "auto_tighten_2x", "enabled": True},
]

# 低风险能力（可自动放行候选）
_LOW_RISK_CAPS = frozenset({
    "web_search", "web_extract", "search", "read", "file_read",
    "glob", "grep", "list_dir", "memory_read", "wiki_read",
    "knowledge_query", "rag_query",
})

# 高危能力（永远不自动放行）
_DANGER_CAPS = frozenset({
    "command", "shell", "bash", "exec", "file_rw", "file_write",
    "write", "delete", "rm", "sudo", "network", "egress", "browser",
    "http", "mcp", "subagent", "spawn",
})


def _rule_enabled(rules: list[dict[str, Any]], key: str, default: bool = True) -> bool:
    for r in rules:
        if r.get("key") == key:
            return bool(r.get("enabled", default))
    return default


async def load_approval_rules() -> list[dict[str, Any]]:
    """从 settings 表读 approval_rules；失败回落默认。"""
    global _RULES_CACHE
    try:
        from sqlalchemy import select

        from backend.database import AsyncSessionLocal
        from backend.models.setting import Setting

        async with AsyncSessionLocal() as session:
            row = (
                await session.execute(
                    select(Setting).where(Setting.key == "approval_rules")
                )
            ).scalar_one_or_none()
        if row is None or row.value is None:
            _RULES_CACHE = list(DEFAULT_RULES)
            return _RULES_CACHE
        val = row.value
        if isinstance(val, str):
            import json
            val = json.loads(val)
        if isinstance(val, list) and val:
            _RULES_CACHE = val
            return val
    except Exception as e:
        logger.debug("load approval_rules failed: %s", e)
    _RULES_CACHE = list(DEFAULT_RULES)
    return _RULES_CACHE


def rule_enabled_sync(key: str, default: bool = True) -> bool:
    """同步读缓存规则（charge_tokens 等零 await 路径）。"""
    rules = _RULES_CACHE if _RULES_CACHE is not None else DEFAULT_RULES
    return _rule_enabled(rules, key, default)


def classify_caps(capabilities: list[str] | tuple[str, ...] | set[str]) -> str:
    """返回 low / high / upgrade。R4：优先 Rust approval_classify。"""
    caps_list = [str(c) for c in capabilities]
    try:
        from backend.kernel import get_kernel

        k = get_kernel()
        if hasattr(k, "_call"):
            # sync rules to rust best-effort
            rules = _RULES_CACHE if _RULES_CACHE is not None else DEFAULT_RULES
            try:
                k._call("approval_set_rules", {"rules": rules})
            except Exception:
                pass
            r = k._call("approval_classify", {"capabilities": caps_list}) or {}
            kind = str(r.get("kind") or "")
            if kind in ("low", "high", "upgrade"):
                return kind
    except Exception as e:
        logger.debug("approval_classify rust skip: %s", e)
    caps = {c.lower() for c in capabilities}
    if any(
        any(d in c for d in _DANGER_CAPS) or c in _DANGER_CAPS
        for c in caps
    ):
        return "high"
    if caps and caps <= _LOW_RISK_CAPS:
        return "low"
    # 部分命中 low、或未知能力 → 视为能力升级
    return "upgrade"


async def should_auto_approve_escalation(
    capabilities: list[str] | tuple[str, ...],
) -> bool:
    """提权是否可自动批准（仅 auto_low_risk + 纯低风险能力）。"""
    rules = await load_approval_rules()
    # push rules + ask rust
    try:
        from backend.kernel import get_kernel

        k = get_kernel()
        if hasattr(k, "_call"):
            k._call("approval_set_rules", {"rules": rules})
            r = k._call(
                "approval_should_auto",
                {"capabilities": [str(c) for c in capabilities]},
            ) or {}
            if "auto_approve" in r:
                return bool(r.get("auto_approve"))
    except Exception as e:
        logger.debug("approval_should_auto rust skip: %s", e)
    if not _rule_enabled(rules, "auto_low_risk", True):
        return False
    kind = classify_caps(capabilities)
    if kind == "high" and _rule_enabled(rules, "review_high_risk", True):
        return False
    if kind == "upgrade" and _rule_enabled(rules, "review_capability_upgrade", True):
        return False
    return kind == "low"


def evolution_requires_review() -> bool:
    """进化建议是否必须人工（硬默认 True；规则关闭也仍强制——双保险）。"""
    return True  # 永不因规则关闭而自动应用

File: crates/takton-kernel/src/approval_rules.rs
//! Approval rules for escalations (R4) — pure policy, no DB.
//! Evolution never auto-applies (hard red line).

use std::collections::HashSet;

use serde_json::{json, Value};

const LOW_RISK: &[&str] = &[
    "web_search",
    "web_extract",
    "search",
    "read",
    "file_read",
    "glob",
    "grep",
    "list_dir",
    "memory_read",
    "wiki_read",
    "knowledge_query",
    "rag_query",
    "ipc_recv",
];

const DANGER: &[&str] = &[
    "command",
    "shell",
    "bash",
    "exec",
    "terminal",
    "file_rw",
    "file_write",
    "write",
    "delete",
    "rm",
    "sudo",
    "network",
    "egress",
    "browser",
    "http",
    "mcp",
    "subagent",
    "spawn",
];

#[derive(Debug, Clone)]
pub struct ApprovalRules {
    pub auto_low_risk: bool,
    pub review_high_risk: bool,
    pub review_capability_upgrade: bool,
    pub review_evolution: bool,
    pub auto_tighten_2x: bool,
}

impl Default for ApprovalRules {
    fn default() -> Self {
        Self {
            auto_low_risk: true,
            review_high_risk: true,
            review_capability_upgrade: true,
            review_evolution: true,
            auto_tighten_2x: true,
        }
    }
}

impl ApprovalRules {
    pub fn from_json_list(rules: &[Value]) -> Self {
        let mut r = Self::default();
        for item in rules {
            let key = item.get("key").and_then(|v| v.as_str()).unwrap_or("");
            let en = item
                .get("enabled")
                .and_then(|v| v.as_bool())
                .unwrap_or(true);
            match key {
                "auto_low_risk" => r.auto_low_risk = en,
                "review_high_risk" => r.review_high_risk = en,
                "review_capability_upgrade" => r.review_capability_upgrade = en,
                "review_evolution" => r.review_evolution = en,
                "auto_tighten_2x" => r.auto_tighten_2x = en,
                _ => {}
            }
        }
        r
    }

    pub fn to_json_list(&self) -> Value {
        json!([
            {"key": "auto_low_risk", "enabled": self.auto_low_risk},
            {"key": "review_high_risk", "enabled": self.review_high_risk, "warn": true},
            {"key": "review_capability_upgrade", "enabled": self.review_capability_upgrade, "warn": true},
            {"key": "review_evolution", "enabled": self.review_evolution, "warn": true},
            {"key": "auto_tighten_2x", "enabled": self.auto_tighten_2x},
        ])
    }
}

/// low | high | upgrade
pub fn classify_caps(capabilities: &[String]) -> &'static str {
    let caps: HashSet<String> = capabilities.iter().map(|c| c.to_lowercase()).collect();
    let danger: HashSet<&str> = DANGER.iter().copied().collect();
    let low: HashSet<&str> = LOW_RISK.iter().copied().collect();
    for c in &caps {
        if danger.contains(c.as_str()) || danger.iter().any(|d| c.contains(d)) {
            return "high";
        }
    }
    if !caps.is_empty() && caps.iter().all(|c| low.contains(c.as_str())) {
        return "low";
    }
    "upgrade"
}

pub fn should_auto_approve(rules: &ApprovalRules, capabilities: &[String]) -> bool {
    if !rules.auto_low_risk {
        return false;
    }
    let kind = classify_caps(capabilities);
    if kind == "high" && rules.review_high_risk {
        return false;
    }
    if kind == "upgrade" && rules.review_capability_upgrade {
        return false;
    }
    kind == "low"
}

/// Evolution always requires human review (double lock).
pub fn evolution_requires_review() -> bool {
    true
}

pub struct ApprovalPolicy {
    rules: ApprovalRules,
}

impl Default for ApprovalPolicy {
    fn default() -> Self {
        Self {
            rules: ApprovalRules::default(),
        }
    }
}

impl ApprovalPolicy {
    pub fn set_rules(&mut self, rules: ApprovalRules) {
        self.rules = rules;
    }

    pub fn set_from_json(&mut self, list: &[Value]) {
        self.rules = ApprovalRules::from_json_list(list);
    }

    pub fn rules_json(&self) -> Value {
        self.rules.to_json_list()
    }

    pub fn classify(&self, capabilities: &[String]) -> Value {
        json!({
            "kind": classify_caps(capabilities),
            "auto_approve": should_auto_approve(&self.rules, capabilities),
            "evolution_requires_review": evolution_requires_review(),
        })
    }

    pub fn should_auto_approve(&self, capabilities: &[String]) -> bool {
        should_auto_approve(&self.rules, capabilities)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn low_auto_high_not() {
        let r = ApprovalRules::default();
        assert!(should_auto_approve(
            &r,
            &["file_read".into(), "grep".into()]
        ));
        assert!(!should_auto_approve(&r, &["terminal".into()]));
        assert!(evolution_requires_review());
    }
}
